Use Resource[T] to implement CiliumNode watcher

pippolo84 opened the pull request:
Use Resource[T] to implement the CiliumNode k8s watcher. This removes the additional informer previously used, reducing the queries to the kube-apiserver and reducing the memory needed for the additional local cache.

Please refer to the individual commits for further details.

Depends on: ~https://github.com/cilium/cilium/pull/29414~

Related: https://github.com/cilium/cilium/issues/28159

pippolo84 committed:
k8s/watchers: Simplify cilium node watcher for local node updates

After the removal of CiliumNodeChain subscription facility, there is no
need to propagate the newest resource version for CiliumNode objects.

Relates: 0af235cec4 ("watchers: Propagate all local CiliumNode updates")
Relates: 9d0c2011f5 ("watchers: remove ciliumnodechain")

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
k8s/watchers: Use Resource[T] to implement CiliumNode watcher

Use Resource[T] to implement the CiliumNode k8s watcher. This removes
the additional informer previously used, reducing the queries to the
kube-apiserver and reducing the memory needed for the additional local
cache.

Since Resource[CiliumNode] underlying informer should be stopped when
the key value store is connected, the Resource[CiliumNode] constructor
is changed in the agent to be marked as releasable.

Differently to the previous watcher, the new one supports the plumbing
of an external context (like the daemon ctx), in order to stop the
watcher goroutines when it is cancelled.

Also, metrics accounting from the watcher has been removed, since the
underlying Resource[CiliumNode] increments the k8s event metrics by
itself.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

lmb commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,153 +5,159 @@ package watchers
+		select {
+		case <-ctx.Done():
+			return
+		case <-kvstore.Client().Disconnected():
```
How is the goroutine spawned above stopped in this case? I'm guessing there is some interaction with Resource[T] but maybe it makes sense to document that.

Maybe it would be easier to understand what is going on if the `<-kvstore.Connected()` part also happened here?  Bascially `cancel` on the context of the goroutine reading from the event stream.

pippolo84 replied:
> How is the goroutine spawned above stopped in this case? I'm guessing there is some interaction with Resource[T] but maybe it makes sense to document that.

AFAIU the channel returned by `kvstore.Client().Disconnected()` will close only if a connection has previously  established and now is terminating:

https://github.com/cilium/cilium/blob/58624f295ad971781f2b0cae3d66608bf2a48f9a/pkg/kvstore/etcd.go#L499-L503

So, in this case the above goroutine should have already been stopped, since the channel returned by `kvstore.Client().Connected()` has already been closed.

> Maybe it would be easier to understand what is going on if the <-kvstore.Connected() part also happened here? Bascially cancel on the context of the goroutine reading from the event stream.

Good point, I've tried that and it seems to make things a bit clearer. Going to push the updated version :+1:

lmb commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,153 +5,159 @@ package watchers
-	// for example, when the key-value store is connected and the local store might
-	// have stale versions of the CiliumNode resource.
-	if k.ciliumNodeStore == nil {
+	// In case this function is called before starting the Events stream from ciliumNodeInit,
```
I find this a bit confusing, since `forceCiliumNodeFromApiServer` below doesn't actually use the Resource store. Maybe it makes sense under which circumstances the bool is true?

pippolo84 replied:
Yep, that's why I added `... and the key-value store is not connected ...` later.
But maybe it is better to move the comment after the `if`, just above the part of the function that relies on the Resource store :+1:

lmb replied:
Yeah LGTM. I think I'd move the `if the key-value store is connected ...` just before `if k.forceCiliumNodeFromApiServer`

pippolo84 replied:
Done!

lmb reviewed (commented):
Amazing, thanks for looking into this!

lmb commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,153 +5,154 @@ package watchers
+			// CiliumNode are performed with a request sent to kube-apiserver directly instead
+			// of relying on an outdated version of the CiliumNode in the Resource[T] cache.
+			k.forceCiliumNodeFromApiServer.Store(true)
+			k.cancelWaitGroupToSyncResources(apiGroup)
```
Just checking: we don't need do to the same cleanup on exiting from `ciliumNodeInit`?

pippolo84 replied:
I think we can avoid that just like all the other watchers, since `ciliumNodeInit` exits only if the agent is shutting down.

aanm commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,153 +5,153 @@ package watchers
-		close(isConnected)
+		// derive another context to signal Events() in case of kvstore connection
+		eventsCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
```
Just FYI that all of these defers will only happen after we exit this for-loop.

pippolo84 replied:
Good point, I changed the code to add a call to cancel under `case <-ctx.Done():` in the first `select`. That is enough to be sure the previous context will always be cancelled without relying on defer :+1:

aanm commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,153 +5,153 @@ package watchers
+						needUpdate = k.onCiliumNodeUpdate(oldObj, event.Object)
+					}
+					if needUpdate {
+						cache[event.Key] = event.Object.DeepCopy()
```
I would really like to avoid this DeepCopy because we are duplicating memory. Would it be possible to only store the `ObjectMeta` instead since those are the only fields that are compared with the new node in the `onCiliumNodeUpdate` function to check if the node has changed?

pippolo84 replied:
Looking at line 116 we are also comparing the old and the new objects with `DeepEqual`.
I think the additional comparisons you are mentioning have been added on top of that because `DeepEqual` ignores annotations (https://github.com/cilium/cilium/pull/25465), but we still have to consider the entire CiliumNode object.

tommyp1ckles commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -129,27 +129,21 @@ func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient client.Clientset, asyncContro
 	}
 	k.ciliumNodeStoreMU.RUnlock()
 
-	if !exists || getFromAPIServer {
```
I'm a little bit confused here, the commit message says that this removes the fallback API call, but we still do a fallback on line: 141 (with the only difference being that we don't attempt fallback if it's not in the local cache store?).

pippolo84 replied:
> (with the only difference being that we don't attempt fallback if it's not in the local cache store?)

That's right, mentioning "fallback" I was referring to that. IOW, with the new implementation, when `forceCiliumNodeFromApiServer` is false (IOW the kvstore is not connected) we check the local store and, if `nodeName` is not found, we immediately return an error without querying the api server too.

tommyp1ckles commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,167 +5,153 @@ package watchers
+
+		select {
+		case <-kvstore.Connected():
+			log.Info("Connected to key-value store, stopping CiliumNode watcher")
```
Would there be any reason to wait on the stop channel to close notifying that the event loop has processed all events?

pippolo84 replied:
I cannot think of any, since AFAIU the current implementation of Resource[T] closes the channel returned by `Events` from the producer side only when the resource is shut down. But in that case, the entire hive is shutting down, so we're gonna end up in the `case <-ctx.Done():`.

giorio94 commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,153 +5,153 @@ package watchers
+			defer close(stop)
 
-		log.Info("Disconnected from key-value store, restarting CiliumNode watcher")
+			events := k.resources.CiliumNode.Events(eventsCtx)
```
AFAICT, there's a notable difference compared to the previous approach. Earlier, we stopped the informer whenever the connection to the kvstore was established, and restarted it only in the event of a disconnection. With this implementation, instead, we would never stop it, regardless of whether the context passed to Events gets canceled earlier (that only breaks the loop, with the informer continuing to run in background until the stop hook of the corresponding resource is executed.

This would defeat the point of watching node updates from the kvstore, and I think it would be worth additional discussion whether this is something that we really want to get rid of at this point, due to the possible scalability implications (especially because CiliumNode CRs can be pretty chatty in case of ENI IPAM mode). Not that the previous implementation was really great from this point of view, as in case of kvstore disconnection it would cause a full relist form k8s, to likely just discard everything immediately afterwards on kvstore reconnection (my understanding is that it was required to break the :chicken::egg: dependency if the kvstore runs in pod network).

The other point is that, if we preserve the synchronization from the kvstore, then the EGW should not watch directly the CiliumNode CRs, as that prevents it from working in the kvstore/clustermesh case (at least as a contributing factor, not necessarily the only one).

/cc @marseel

pippolo84 replied:
I think this PR is still an improvement compared to the previous implementation. From the agent point of view, two different informers were active at the same time for the same resource. One of the informer, namely the one started from Resource[T] used in the Egress Gateway manager, would never have been stopped in the previous version too.

With the proposed changes, it is true that cancelling the context passed to `Events` won't stop the Resource[T] informer either, but that is expected, since it is still needed to pump the updates to the Egress Gateway manager.

If we decide to avoid depending on the Resource[CiliumNode] (or unsubscribe from it dynamically) in the EG manager, then what we probably need is the ability of Resource[T] to stop the underlying informer when there are no subscribers anymore.

giorio94 replied:
> From the agent point of view, two different informers were active at the same time for the same resource. One of the informer, namely the one started from Resource[T] used in the Egress Gateway manager, would never have been stopped in the previous version too.

Hmm, that's true, but only if the Egress Gateway is enabled, and [~~kvstore is not enabled (the two are not compatible anyway)~~](https://github.com/cilium/cilium/pull/29222#discussion_r1401895952). When kvstore is enabled, instead, this PR is introducing a "regression"  there, because the expectation that the informer gets stopped after connecting to the kvstore does no longer hold true.

I personally don't have a strong opinion on what approach is best, each has its own advantages and disadvantages. Although my understanding is that the Egress Gateway should not watch the CiliumNodes in the first place, but use an higher-level abstraction to be compatible both with CRD and kvstore mode.

giorio94 replied:
And the same behavioral change would be also caused for CiliumEndpoints by [this other commit](https://github.com/cilium/cilium/pull/29249/commits/6386f9490d168cdc9572f37721cbbb731788c497) (and the equivalent one for CiliumEndpointSlices).

pippolo84 replied:
The fact that this issue may affect more than one resource types makes me think that it should be addressed in the underlying package. Let's see if adding this to Resource[T]:

> what we probably need is the ability of Resource[T] to stop the underlying informer when there are no subscribers anymore.

is feasible. I'm gonna try that in another PR :+1:

giorio94 replied:
> what we probably need is the ability of Resource[T] to stop the underlying informer when there are no subscribers anymore.

I'm not really sure whether this is really feasible when used in combination with the Store API, as after stopping it would contain stale data only. And that could break other (unaware) consumers in many unexpected ways.

More in general, maybe this thread could be a good topic for discussion during today's community meeting. WDYT?

pippolo84 replied:
> > what we probably need is the ability of Resource[T] to stop the underlying informer when there are no subscribers anymore.
> 
> I'm not really sure whether this is really feasible when used in combination with the Store API, as after stopping it would contain stale data only. And that could break other (unaware) consumers in many unexpected ways.

Actually the Store API marks the informer as needed and block the caller till the cache is synced again. I was thinking about keeping this behavior while introducing the change above.

> More in general, maybe this thread could be a good topic for discussion during today's community meeting. WDYT?

That would be great, super happy to discuss this in more detail. :+1:

giorio94 replied:
> Actually the Store API marks the informer as needed and block the caller till the cache is synced again. I was thinking about keeping this behavior while introducing the change above.

Yep, my point is that stopping the informer when events are no longer observed, but keeping it if the anyone gathered the store, would defeat the whole idea of stopping the informer (as for instance we always retrieve the node store IIRC). And it would also introduce more uncertainty on whether a given resource is actually being watched or not (as the stopping conditions would also play a role, in addition to the starting ones).

> That would be great, super happy to discuss this in more detail. 👍

:+1:  I'll add the topic to the agenda

giorio94 replied:
> the Egress Gateway is enabled, and kvstore is not enabled (the two are not compatible anyway)

Most likely I was wrong about this incompatibility, because the CiliumNode CRs are present (and updated) regardless of the kvstore mode. While the point still holds in case of Cluster Mesh.

marseel replied:
tldr at the end

> I'm not really sure whether this is really feasible when used in combination with the Store API, as after stopping it would contain stale data only.

I think it would make sense to "force" consumers to comply with this requirement - consumers should depend on events to trigger updates and not rely on the store only  - so if they are consumers, they should have a channel for receiving events (and if they close it they should know that store will be obsolete) etc and respect this assumption (I understand this might not be an easy change though).

I didn't have a strong preference before if we should/shouldn't stop informers when switching from CRD to kvstore in this case. The reasoning is that:
- if we close informer, then once kvstore somehow gets broken, all of the agents will spam the k8s control plane simultaneously, which is not great and will result in degraded performance for a long time (I am guessing possibly tens of minutes) - still not as bad contributor to overall load as Endpoints
- if we don't close the informer, we still put some pressure on k8s control plane, probably insignificant in most of the cases (unless we have ENI mode as mentioned before for the Node CRD case)
- also, one side effect of having watches opened and not closing informer will be that update requests to Node CRDs will get slower (k8s takes into account the number of watches opened to specific resource in the context of throttling update requests, reducing throughput of updates up to 10x). 

This makes me think that we should have a general guideline on when to close/not close informers in the case of kvstore as it's super non-obvious. 
I would suggest:
- Frequent Node updates in ENI mode is a separate issue that should be addressed regardless of this PR. I think the best solution would be to fix ENI issue + keep the watch open. But until we fix the ENI issue, closing the informer would make it at least as good as before. 
- We should do some retries after disconnecting from kvstore before trying to get data from CRD to protect the k8s control plane (I will open a separate issue for that) - this is especially important for other resources like Endpoints/Pods (we watch all Pods when Endpoint CRD is disabled, that's a separate issue) which have much higher cardinality and churn.
- It should be fine to keep watches for all other resources - local Endpoints, Identities, Nodes, ... in general etc So basically only avoiding Endpoints/Pods.

So now the question becomes, what's the purpose of having kvstore:
- Offload Endpoint load/churn to kvstore in k8s clusters (still we have the same issue with switching between CRD and kvstore for this resource, I think we would need to figure out a better solution for it)
- Cilium deployments without a k8s control plane
- Expose data in case of some of the clustermesh deployments - which hopefully we would like to get rid of in long-term
- kvstore identity allocation mode - I don't think it brings many benefits in k8s clusters(and doesn't allow to use "regular" clustermesh control plane, which forces users to stick with the previous point), but is required in cilium deployments without a k8s control plane
- allows to disable CiliumEndpoint CRD in k8s clusters - I believe this actually brings more harm than benefits currently as it forces agents to list/watch all pods until connecting to kvstore (similar issue to the first point). 
- ... something else I missed?

tldr
- For this PR, closing informer once connected to kvstore is fine and as good as it was before. It would be also fine to not close the informer if we fix the ENI issue. 
- We should improve retries of disconnects from kvstore to reduce the load on the k8s control plane in case of Etcd failures
- long-term offloading the k8s control plane with Nodes/Identities churn (after fixing the ENI issue) shouldn't be a priority for kvstore deployments and we should probably focus on better handling of switches between CRD/kvstore in case of Endpoints (and Pods when Endpoints CRD is disabled).
- we **really** need some good abstraction for fetching and storing data from/to CRDs/kvstore/clustermesh (stateDB?)

pippolo84 replied:
Solved introducing a releasable Resource[CiliumNode] (see https://github.com/cilium/cilium/pull/29414).

pippolo84 committed:
operator/k8s: Use a separate constructor for CiliumNode

Agent constructor for Resource[CiliumNode] should set the
WithStoppableInformer option to be able to stop the informer in case of
a key value store connection. This will be done in a subsequent commit.

Since this is not needed in the operator, the commit adds a separate
constructor for the Resource[CiliumNode] used in the operator.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

giorio94 commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -129,27 +129,21 @@ func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient client.Clientset, asyncContro
+	// have stale versions of the CiliumNode resource.
 	if k.ciliumNodeStore == nil {
-		getFromAPIServer = true
+		return k.clientset.CiliumV2().CiliumNodes().Get(ctx, nodeName, v1.GetOptions{})
```
The lock should be released before this line (this is a nit, as it only applies to the second commit, and it is then fixed in the forth).

pippolo84 replied:
I'm gonna remove the second commit to address André's feedback.

giorio94 commented on operator/k8s/resource_ctors.go:
```diff
@@ -57,3 +57,16 @@ func CiliumEndpointSliceResource(lc hive.Lifecycle, cs client.Clientset, opts ..
 	return resource.New[*cilium_api_v2alpha1.CiliumEndpointSlice](lc, lw, resource.WithMetric("CiliumEndpointSlice")), nil
 }
+
+func CiliumNodeResource(lc hive.Lifecycle, cs client.Clientset, opts ...func(*metav1.ListOptions)) (resource.Resource[*cilium_api_v2.CiliumNode], error) {
```
> Since this is not needed in the operator, the commit adds a separate
constructor for the Resource[CiliumNode] used in the operator.

The same also applies in the clustermesh-apiserver case.

pippolo84 replied:
@giorio94 I've added a similar commit to have a separate CN resource ctors for clustermesh-apiserver, too :+1:

giorio94 replied:
Thanks!

aanm commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -129,27 +129,21 @@ func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient client.Clientset, asyncContro
-		getFromAPIServer = true
+		return k.clientset.CiliumV2().CiliumNodes().Get(ctx, nodeName, v1.GetOptions{})
 	} else {
 		nodeInterface, exists, err = k.ciliumNodeStore.GetByKey(nodeName)
```
Why aren't we still failing back to the kubeapiserver in case it doesn't exist on the local store?

pippolo84 replied:
My understanding is that the fallback was needed for the Egress Gateway manager, but now that it uses Resource[CiliumNode] it should not be needed anymore (see [this](https://github.com/cilium/cilium/pull/29222/commits/24736e1bc9767a7279667806f065a23c817a9f30) commit message for the references)

pippolo84 replied:
So, looking again at the commit, I'm afraid I was confused with a related previous change to the CEP watcher. I'll remove this commit, but I'll keep this behavior (no fallback to the k8s apiserver when local store is not nil) in the last commit. That's because in the last commit we move the watcher to rely on Resource[T] and Resource[T] blocks until Store is fully synchronized: https://github.com/cilium/cilium/blob/fa0037638534635471c8e45583d78f87ad78b3e1/pkg/k8s/resource/resource.go#L262-L269

giorio94 commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,153 +5,162 @@ package watchers
+func (k *K8sWatcher) onCiliumNodeUpdate(oldNode, newNode *cilium_v2.CiliumNode) bool {
+	// Comparing Annotations here since wg-pub-key annotation is used to exchange rotated WireGuard keys.
+	if oldNode.DeepEqual(newNode) &&
+		comparator.MapStringEquals(oldNode.ObjectMeta.Labels, newNode.ObjectMeta.Labels) &&
```
Nit. Could be replaced with `maps.Equal` from the standard library.

giorio94 commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,153 +5,162 @@ package watchers
-// local store is not initialized then it will fallback retrieving the node
-// from kube-apiserver.
+func (k *K8sWatcher) onCiliumNodeDelete(ciliumNode *cilium_v2.CiliumNode) {
+	n := types.ParseCiliumNode(ciliumNode)
```
Nit. You may want to skip the event for the local node, for consistency.

giorio94 commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,153 +5,162 @@ package watchers
+			return
+		case <-kvstore.Client().Disconnected():
+			log.Info("Disconnected from key-value store, restarting CiliumNode watcher")
+		}
```
I'd suggest adding a wait group to make sure that goroutines terminated before restarting a new loop iteration. Although extremely unlikely in practice (the kvstore should disconnect immediately after connecting), running two copies in parallel could cause subtle race conditions.

giorio94 reviewed (approved):
/lgtm, thanks!

Just a couple of nits inline.

pippolo84 committed:
clustermesh-apiserver: Use a separate constructor for CiliumNode

Agent constructor for Resource[CiliumNode] should set the
WithStoppableInformer option to be able to stop the informer in case of
a key value store connection. This will be done in a subsequent commit.

Since this is not needed in the clustermesh-apiserver, the commit adds a
separate constructor for the Resource[CiliumNode] used in the
clustermesh-apiserver.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,167 +5,174 @@ package watchers
+
+func (k *K8sWatcher) onCiliumNodeUpdate(oldNode, newNode *cilium_v2.CiliumNode) bool {
+	// Comparing Annotations here since wg-pub-key annotation is used to exchange rotated WireGuard keys.
+	if oldNode.DeepEqual(newNode) &&
```
Doesn't `DeepEqual` already compare the `Labels` and `Annotations`? 

Also wouldn't it make more sense to compare `ResourceVersion` rather than `DeepEqual`? Doing `DeepEqual` perhaps makes sense with the slim objects that only have subset of the fields, but here I would expect that change in `ResourceVersion` will imply that one of the fields have changed.

OTOH, perhaps let's not add a last minute optimization like this as it can go wrong in multiple ways :)

pippolo84 replied:
> Doesn't `DeepEqual` already compare the `Labels` and `Annotations`?

Nope, `DeepEqual` does not compare the `Annotations` (see https://github.com/cilium/cilium/pull/25465/files)

> OTOH, perhaps let's not add a last minute optimization like this as it can go wrong in multiple ways :)

Happy to try that in a follow up PR, though :+1:

joamaki commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,167 +5,174 @@ package watchers
+					synced.Store(true)
+				case resource.Upsert:
+					var needUpdate bool
+					oldObj, ok := cache[event.Key]
```
Is this "insert" stuff really needed? `onCiliumNodeUpdate` calls `onCiliumNodeInsert`. could we only have the update and get rid of the `cache` thing?

pippolo84 replied:
Without the cache we couldn't filter out "update" events for which `oldNode.DeepEqual(newNode) && maps.Equal(...)` is true, thus calling `types.ParseCiliumNode` and `k.nodeDiscoverManager.NodeUpdated(n)` more times than needed.
TBH I'm not sure about the performance impact of that, but being a refactoring I chose to keep the old behavior.

joamaki commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,167 +5,174 @@ package watchers
+		return false
+	}
+	n := types.ParseCiliumNode(ciliumNode)
+	k.nodeDiscoverManager.NodeUpdated(n)
```
Since this seems to be the only thing happening here wouldn't it be easy enough to make `NodeDiscovery` subscribe to `Resource[*CiliumNode]` and just remove `pkg/k8s/watchers/cilium_node.go` altogether? 

Only complication perhaps is around `blockWaitGroupToSyncResources`.. e.g. something might be waiting on the "caches synced" to be sure that `NodeUpdate` has been called for all `List`d nodes.
EDIT: Second complication is that `NodeDiscovery` would need to make sure to stop the subscription when kvstore connects. I think it makes sense to leave `cilium_node.go` for now and go for a proper solution that deals with the kvstore complication in a principled way instead (which would be to never expose `CiliumNode` CRD directly)

tl;dr: Nevermind this comment.

giorio94 replied:
> make NodeDiscovery subscribe to Resource[*CiliumNode]

For completeness, the node discovery module only needs to subscribe to Resource[LocalCiliumNode], which doesn't need to be stopped.

tommyp1ckles commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -252,8 +254,8 @@ type K8sWatcher struct {
 
-	ciliumNodeStoreMU lock.RWMutex
-	ciliumNodeStore   cache.Store
+	ciliumNodeStore   resource.Store[*cilium_v2.CiliumNode]
```
(optional)💡 based on the usage it seems we could use an atomic pointer here, one less mutex lock to think about.

pippolo84 replied:
That's right, I've updated the code to use an `atomic.Pointer` :+1:

tommyp1ckles commented on operator/k8s/resource_ctors.go:
```diff
@@ -57,3 +57,16 @@ func CiliumEndpointSliceResource(lc hive.Lifecycle, cs client.Clientset, opts ..
 	return resource.New[*cilium_api_v2alpha1.CiliumEndpointSlice](lc, lw, resource.WithMetric("CiliumEndpointSlice")), nil
 }
+
+func CiliumNodeResource(lc hive.Lifecycle, cs client.Clientset, opts ...func(*metav1.ListOptions)) (resource.Resource[*cilium_api_v2.CiliumNode], error) {
```
nit: do we need to return an error here, the `cell.Provide(ctors...)` can take either?

pippolo84 replied:
Currently, there is no case where the ctor can fail, but if an error occurs, the hive calls log.Fatal with a message like:

`failed to build resource.Resource[*github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2.CiliumNode]: received non-nil error from function \"github.com/cilium/cilium/operator/k8s\".CiliumNodeResource`

tommyp1ckles commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -5,167 +5,167 @@ package watchers
+			store, err := k.resources.CiliumNode.Store(subCtx)
+			if err != nil {
+				if !errors.Is(err, context.Canceled) {
+					log.WithError(err).Warning("unable to retrieve CiliumNode local store, going to query kube-apiserver directly")
```
Bit confused about this, where does the directly querying apiserver take place in this scenario?

pippolo84 replied:
At line 157. This might happen if `Store` fails or:

- at startup, if `GetCiliumNode` is called before the store is synchronized
- if kvstore is connected and the informer is stopped

tommyp1ckles reviewed (approved):
Looks good, thanks @pippolo84
